In [ ]:
#Basics of Functions:

In [ ]:
#Q1. What is a function in Python, and why is it used?

1) Python Functions is a block of statements that return the specific task.

2) That means that a function is a piece of code written to carry out a specified task.

3) In programming, functions allow us to repeat a sequence of code without having to write the code over again. 

4) This allows us to simplify a complex action by using only one command, just like in our criss-cross applesauce example.

5) Syntax

  def function_name(parameters):
  
   #Function body
   
  return output

In [ ]:
#Q2.How do you define a function in Python? Provide an example.

 Syntax

  def function_name(parameters):
   #Function body
  return output
1) def is the keyword used to define a function.

2) Function_name is the name of the function.

3) Parameters are the arguments that the function takes.

4) Docstring is an optional string that describes the function.

5) Function body is the block of code that the function executes.

6) Return is a keyword used to return a value from the function.

7) Here is an example of a function definition:

    def add_numbers(a, b):
    
  """Returns the sum of two numbers."""
  
  return a + b
  
  result = add_numbers(1, 2)

Output:3

In [ ]:
#Q3. Explain the difference between a function definition and a function call.

Function in Python

In Python, a function is a block of code that performs a specific task and may or may not return a value. A function is defined using the def keyword, followed by the function name and a set of parentheses that may include parameters. Here is an example of a function definition in Python:

def greet(name):

print("Hello, " + name)

This function, called greet, takes a single parameter called name and prints a greeting message using the value of name.

Function call in Python

A function call is an expression that invokes a function. To call a function in Python, you simply write the function name followed by a set of parentheses that may include arguments. Here is an example of a function call in Python:

greet("John")

This function call invokes the greet function with the argument "John", which will cause it to print the message "Hello, John".

In [ ]:
#Q4. Write a Python program that defines a function to calculate the sum of two numbers and then calls the function.

In [1]:
def add_two_num(a,b):
    sum=a+b;
    return sum; 
num1=int(input("Input the first number : "))
num2=int(input("Input the second number  :"))
print("The sum of given two numbers is",add_two_num(num1,num2))

Input the first number :  34
Input the second number  : 45


The sum of given two numbers is 79


1). Simply defined the function that accepts two arguments. 

2). Used the + operator to add those two numbers and return the value stored in the sum variable.

3). We used the input() function to accept the input from the user.

In [ ]:
#Q5. What is a function signature, and what information does it typically include?

1). The signature indicates the names and types of the input arguments, and (with type annotations) the type of the returned result(s) of a function or method.

2). A function signature is its declaration, parameters, and return type.

3). A signature can include: parameters and their types. a return value and type. exceptions that might be thrown or passed back.

In [ ]:
#Q6. Create a Python function that takes two arguments and returns their product.

In [3]:
def mul_two_num(a,b):
    mul=a*b;
    return mul; 
num1=int(input("Input the first number : "))
num2=int(input("Input the second number  :"))
print("The sum of given two numbers is",mul_two_num(num1,num2))

Input the first number :  34
Input the second number  : 45


The sum of given two numbers is 1530


1). Simply defined the function that accepts two arguments.

2). Used the * operator to multiply those two numbers and return the value stored in the mul variable.

3). We used the input() function to accept the input from the user.

In [ ]:
#Function Parameters and Arguments:

In [ ]:
#Q1. Explain the concepts of formal parameters and actual arguments in Python functions.

Formal parameter — the identifier used in a method to stand for the value that is passed into the method by a caller.
For example, amount is a formal parameter of processDeposit

Actual parameter — the actual value that is passed into the method by a caller.
For example, the 200 used when processDeposit is called is an actual parameter.
actual parameters are often called arguments

In [ ]:
#Q2.Write a Python program that defines a function with default argument values.

In [6]:
def student(firstname="Mark", lastname ='Zuckerberg', standard ='Fifth'):
    print(firstname, lastname, 'studies in', standard, 'Standard')

firstname=str(input("Input First name:"))
lastname =str(input("Input Last name:"))
standard =str(input("Input Standard:"))
student(firstname,lastname,standard)
student()

Input First name: Vikas
Input Last name: Ranjan
Input Standard: Ninth


Vikas Ranjan studies in Ninth Standard
Mark Zuckerberg studies in Fifth Standard


1). In the case of passing the keyword arguments, the order of arguments is important.

2). There should be only one value for one parameter.

3). The passed keyword name should match with the actual keyword name.

4). In the case of calling a function containing non-keyword arguments, the order is important.

In [ ]:
#Q3. How do you use keyword arguments in Python function calls? Provide an example.

In [7]:
def display_info(first_name, last_name):
    print('First Name:', first_name)
    print('Last Name:', last_name)

display_info(last_name = 'Ranjan', first_name = 'Vikas')

First Name: Vikas
Last Name: Ranjan


In keyword arguments, arguments are assigned based on the name of arguments. 

Here, we have assigned names to arguments during the function call.

Hence, first_name in the function call is assigned to first_name in the function definition. Similarly, last_name in the function call is assigned to last_name in the function definition.

In such scenarios, the position of arguments doesn't matter.

In [ ]:
#Q4. Create a Python function that accepts a variable number of arguments and calculates their sum.

In [8]:
def sum_all(*args): 
    t=() 
    s=0 
    for i in args: 
        s=s+i 
    print(s) 

sum_all(11,22,33,44,55,66,77,88,99)

495


1). Define the function with argument as *args and define 2 variables s &t.

2). Take a for loop and run it for i in args and store the sum in s.

3). Finally print s and programe over.

In [ ]:
#Q5. What is the purpose of the `*args` and `**kwargs` syntax in function parameter lists?

The *args and **kwargs syntax in function parameter lists allows a function to accept a variable number of arguments.

*args is a tuple that will contain any positional arguments that are passed to the function beyond the number of required arguments. For example, if a function has two required arguments and you pass it three arguments, the third argument will be stored in *args.

**kwargs is a dictionary that will contain any keyword arguments that are passed to the function. For example, if a function has two required arguments and you pass it three arguments, the third argument will be stored in kwargs.

The *args and **kwargs syntax can be useful for writing functions that are more flexible and can be used with a wider range of arguments.

In [ ]:
#Return Values and Scoping:

In [ ]:
#Q1. Describe the role of the `return` statement in Python functions and provide examples.

1). A return statement is used to end the execution of the function call and “returns” the result to the caller.

2). The statements after the return statements are not executed. If the return statement is without any expression, then the special value None is returned.

3) A return statement is overall used to invoke a function so that the passed statements can be executed.

4) Syntax:
def fun():
    
    statements
    
    .
    .
    
    return [expression]

5). Example :

In [2]:
def percentage(total, obtained):
    per=(obtained/total)*100
    return per

total=500
obtained=int(input("Enter your total marks :"))
print("The percentange you got in this exam : ",percentage(total, obtained))

Enter your total marks : 354


The percentange you got in this exam :  70.8


In [ ]:
#Q2. Explain the concept of variable scope in Python, including local and global variables.

1).In Python, the scope of a variable determines where in the program the variable can be accessed.

2).There are two types of scopes in Python: local and global.

3).Local variables are declared inside a function and can only be accessed within that function. 

4).Global variables are declared outside of any function and can be accessed from anywhere in the program.

5).Here is an example of local and global variables:

In [3]:
name="Vikas" #Global variable 
def print_name():
    name="Virat"
    print ("Hello",name)
    
print_name()
print ("Hello",name)

Hello Virat
Hello Vikas


In [ ]:
#Q3. Write a Python program that demonstrates the use of global variables within functions.

Variables that are created outside of a function are known as global variables.

Global variables can be used by everyone, both inside of functions and outside.

If you create a variable with the same name inside a function, this variable will be local, and can only be used inside the function. 

The global variable with the same name will remain as it was, global and with the original value.

In [ ]:
#Q4.Create a Python function that calculates the factorial of a number and returns it.

In [7]:
def fact(n):
    i=1
    for m in range (1,n+1):
        i=i*m
    return i

n=int(input("Enter your desired number to find factorial :"))
print("The factorial of your desired number is :",fact(n))

Enter your desired number to find factorial : 5


The factorial of your desired number is : 120


1). Take input from user for finding factorial of that number.

2). Define a function fact and take another variable i and initialize with 1.

3). Take for loop with a range 1 to n+1 and store the product of i and m.

4). Finally return i and print it , end of program.

In [ ]:
#Q5.How can you access variables defined outside a function from within the function?

To access variables defined outside a function from within the function, you can use the global keyword. 

The syntax is:
    
    global variable_name

This will tell Python that you are referring to the global variable, and not a local variable with the same name.

Example :

In [1]:
x = 10

def my_function():
  global x
  print(x)

my_function()

10


In [ ]:
#Lambda Functions and Higher-Order Functions:

In [ ]:
#Q1.What are lambda functions in Python, and when are they typically used?

Lambda functions are similar to user-defined functions but without a name. They're commonly referred to as anonymous functions.

Lambda functions are efficient whenever you want to create a function that will only contain simple expressions – that is, expressions that are usually a single line of a statement.

They're also useful when you want to use the function once.

Syntax :

lambda argument(s) : expression

In [ ]:
#Q2.Write a Python program that uses lambda functions to sort a list of tuples based on the second element

In [2]:

subject_marks = [('English', 88), ('Science', 90), ('Maths', 97), ('Social sciences', 82)]


print("Original list of tuples:")
print(subject_marks)


subject_marks.sort(key=lambda x: x[1])

print("\nSorting the List of Tuples:")
print(subject_marks) 

Original list of tuples:
[('English', 88), ('Science', 90), ('Maths', 97), ('Social sciences', 82)]

Sorting the List of Tuples:
[('Social sciences', 82), ('English', 88), ('Science', 90), ('Maths', 97)]


1). Create a list of tuples named 'subject_marks', each tuple containing a subject and its corresponding marks

2). Display the original list of tuples to the console

3). Sort the 'subject_marks' list of tuples based on the second element of each tuple (the marks),

4). Use a lambda function as the sorting key to extract the second element

5). Display the sorted list of tuples to the console

In [ ]:
#Q3 Explain the concept of higher-order functions in Python, and provide an example.

A function that takes another function as an argument, returns a function as a result, or both, is known as a higher-order function.

These functions are a fundamental concept in functional programming, which is a programming paradigm that treats computation as the evaluation of mathematical functions.

In Python, there are several built-in functions that are considered higher-order functions. These include:
                        
                        map(), which applies a function to each element of an iterable and returns a new iterable with the results.
                        
                        filter(), which returns an iterable containing the elements of an iterable that satisfy a given condition.
                        
                        reduce(), which applies a function to two elements of an iterable and returns a single value. 
                        This function is often used to perform calculations such as summation or multiplication.
                        
Here is an example of a higher-order function in Python:

In [3]:
def my_map(func, iterable):
  result = []
  for element in iterable:
    result.append(func(element))
  return result

def square(x):
  return x * x

numbers = [1, 2, 3, 4]
squared_numbers = my_map(square, numbers)

print(squared_numbers)

[1, 4, 9, 16]


In [ ]:
#Q4. Create a Python function that takes a list of numbers and a function as arguments, applying the function to each element in the list.

In [1]:
def apply_to_all(numbers, func):
    results = []
    for number in numbers:
        result = func(number)
        results.append(result)
    return results


numbers = [1, 2, 3]
squared_numbers = apply_to_all(numbers, lambda x: x * x)
print(squared_numbers)

[1, 4, 9]


 Applies a function to all elements of a list.
    
  Args:
    numbers: A list of numbers.
    func: A function that takes a number as an argument and returns a number.

  Returns:
    A list of the results of applying the function to each element of the list.


In [ ]:
#Built-in Functions:

In [ ]:
#Q1.Describe the role of built-in functions like `len()`, `max()`, and `min()` in Python.

len(): This function returns the length of an object. For example, the length of a string is the number of characters in the string. The length of a list is the number of elements in the list.

max(): This function returns the largest element in an iterable. For example, the largest element in a list of numbers is the number that is the greatest than all the other numbers in the list.

min(): This function returns the smallest element in an iterable. For example, the smallest element in a list of numbers is the number that is the least than all the other numbers in the list.

In [ ]:
#Q2. Write a Python program that uses the `map()` function to apply a function to each element of a list.

In [ ]:

 
# Return double of n
def addition(n):
    return n + n

numbers = (1, 2, 3, 4)
result = map(addition, numbers)
print(list(result))

1) Python program to demonstrate working of map.

2) Create a function and  Return double of n.

3) Take numbers and call the function using map function .

4) Store it in a variable result and at last print it in form of list.

In [1]:
#Q3. How does the `filter()` function work in Python, and when would you use it?

Filter() is a built-in function in Python. The filter function can be applied to an iterable such as a list or a dictionary and create a new iterator. This new iterator can filter out certain specific elements based on the condition that you provide very efficiently. 

The syntax for the filter() function in Python is - 

filter(function, iterable)

Examples of Filter in Python

In [2]:
def check(letter):
    list_of_vowels = ['a', 'e', 'i', 'o', 'u']
    if letter in list_of_vowels:
        return True
    else:
        return False
letters = ['u', 'a', 'q', 'c', 'i', 'd', 'z', 'p', 'e']
filtered_object = filter(check, letters)
print("The type of returned object is: ", type(filtered_object))
filtered_list = list(filtered_object)
print("The list of vowels is: ", filtered_list)

The type of returned object is:  <class 'filter'>
The list of vowels is:  ['u', 'a', 'i', 'e']


In [ ]:
#Q4. Create a Python program that uses the `reduce()` function to find the product of all elements in a list.

In [4]:
from functools import reduce  
my_list = [1, 2, 3, 4, 5]  
product = reduce(lambda x, y: x * y, my_list)   
print(f"Product = {product}") 

Product = 120


Creating a list and initialize the value

Calculating the product of the numbers in my_list using reduce and lambda functions together  

Store it in a variable & Print output  

Output : Product = 120  

In [ ]:
#Function Documentation and Best Practices:

In [ ]:
#Q1. Explain the purpose of docstrings in Python functions and how to write them.

A Python docstring is a string used to document a Python module, class, function or method, so programmers can understand what it does without having to read the details of the implementation.

Also, it is a common practice to generate online (html) documentation automatically from docstrings.

How to write docstrings in Python functions is as follow:

A docstring is a string literal that occurs as the first statement in a module, function, class, or method definition.

Docstrings are used to provide documentation for Python modules, classes, and methods, and are typically written in a specialized syntax called "reStructuredText" that is used to create formatted documentation.

Docstrings should be descriptive and should explain what the function does, what parameters it takes, and what it returns.

Docstrings should also be consistent in style and format.

There are two main types of docstrings: single-line and multi-line.

Single-line docstrings are typically used for simple functions that can be explained in a single line.

Multi-line docstrings are typically used for more complex functions that require more explanation.

Here is an example of a single-line docstring:

In [5]:
def add_two_numbers(a, b):
  """Returns the sum of two numbers."""
  return a + b

a=int(input("First Number :"))
b=int(input("Second Number :"))
add_two_numbers(a, b)

First Number : 34
Second Number : 45


79

In [6]:
#Q2. Describe some best practices for naming functions and variables in Python, including naming conventions and guidelines.

Python variables support three popular naming conventions:

Camel case, where the first letter is lowercase and each subsequent word in a name begins with an uppercase letter.

Snake case, where separate words are separated by underscores.

Pascal case, where each word in a variable name begins with an uppercase letter.

Python does not have an official naming convention, so you are free to choose between these.